### PR TITLE
linter: Move Flags handling closer to main

### PR DESCRIPTION
**Pull request by tgorochowik**
Cleaning up the flag accesses as promised in #237

### CODE REVIEW

**Comment by tgorochowik** [comment]
I added the changes you asked for and added a generic unit test.

File: verilog/analysis/verilog_linter_configuration.h [hunk 1]
```diff
@@ -130,6 +130,19 @@ struct ProjectPolicy {
...
+  // strings, ints, bools, and unprocessed values from flags, no other derived
+  // information. Reasonable default values may be specified here for each
+  // member
+  const RuleSet& ruleset;
+  const RuleBundle& rules;
+  absl::string_view config_file;
+  bool config_file_is_custom;
+  bool rules_config_search;
+  absl::string_view waivers_file;
+  absl::string_view linting_start_file;
```
Comment (fangism) [inline verilog/analysis/verilog_linter_configuration.h:143]: Add a brief description for each of these members.

File: verilog/analysis/verilog_linter_configuration_test.cc [hunk 2]
```diff
@@ -775,5 +775,22 @@ TEST(RuleBundleTest, ParseRuleBundleSkipComments) {
   }
 }
 
+// ConfigureFromFlags Tests
+TEST(ConifigureFromFlagsTest, Basic) {
```
Comment (fangism) [inline verilog/analysis/verilog_linter_configuration_test.cc:779]: `ConfigureFromFlagsTest` (spelling)

File: verilog/analysis/verilog_linter_configuration_test.cc [hunk 3]
```diff
@@ -775,5 +775,22 @@ TEST(RuleBundleTest, ParseRuleBundleSkipComments) {
...
+      false,
+      ".",
+      "filename"};
+
+  LinterConfiguration config;
+  auto status = config.ConfigureFromFlags(flags);
+
+  EXPECT_TRUE(status.ok());
+}
+
```
Comment (fangism) [inline verilog/analysis/verilog_linter_configuration_test.cc:794]: Good so far.  Can you test to cover some of the other behavior/code paths in `ConfigureFromFlags`?

Comment (tgorochowik) [reply verilog/analysis/verilog_linter_configuration_test.cc:845]: Thanks for the review! I've added some more tests. I don't think we can add unit tests for all of what is being done there without extra refactoring of the code

Comment (fangism) [reply verilog/analysis/verilog_linter_configuration_test.cc:845]: Can you leave a `//TODO:` note around here with what could/should be covered in the next refactoring?

Comment (tgorochowik) [reply verilog/analysis/verilog_linter_configuration_test.cc:848]: Sure added a comment explaining what I meant - basically that if we also read the configuration files in `main()` we will be able to test the configuration form the content of the configuration files too.

Comment (hzeller) [reply verilog/analysis/verilog_linter_configuration_test.cc:848]: Isn't the result of reading a configuration file just a rule-set ? Passing around a LinterConfiguration should contain all that. 
If we want to extend the notion of the _flags_ values being set, we should not call this not LinterConfigurationFlags, but LinterConfigurationOptions, so that it is less tied to the notion of flags (also that is a common way to name such things).

Comment (hzeller) [reply verilog/analysis/verilog_linter_configuration_test.cc:848]: (for testing reading from files, there is a [ScopedTestFile](https://github.com/google/verible/blob/master/common/util/file_util.h#L59) when this becomes important)

File: verilog/analysis/verilog_linter_configuration.h [hunk 4]
```diff
@@ -130,6 +130,30 @@ struct ProjectPolicy {
...
+  // information. Reasonable default values may be specified here for each
+  // member
+
+  // The base set of rules used by linter
+  const RuleSet& ruleset;
+  // Bundle of rules to enable/disable in addition to the base set
+  const RuleBundle& rules;
+  // Path to a file with extra linter configuration, applied on top of the
+  // base 'ruleset' and extra 'rules'
+  absl::string_view config_file;
```
Comment (hzeller) [inline verilog/analysis/verilog_linter_configuration.h:144]: I would not store string_views for the config file, linting_start_file and waivers_file.
The struct is meant to be passed around with the configuration, but we don't know anything about the lifetime of the strings we set it to. So all these should be std::string.

File: verilog/analysis/verilog_linter_configuration.h [hunk 5]
```diff
@@ -130,6 +130,30 @@ struct ProjectPolicy {
   std::string ListPathGlobs() const;
 };
 
+struct LinterConfigurationFlags {
+  // strings, ints, bools, and unprocessed values from flags, no other derived
+  // information. Reasonable default values may be specified here for each
+  // member
+
+  // The base set of rules used by linter
+  const RuleSet& ruleset;
```
Comment (hzeller) [inline verilog/analysis/verilog_linter_configuration.h:139]: This essentially is a pointer to a rule-set and given that the struct is meant to be passed around self-contained, I think this should be a  copy of a rule-set. 

If this is not meant to be an instance but in fact a reference (e.g. because you want to build trees of rules-sets), I'd make it a const pointer, so that the lifetime-implications are very clear (and they shoudl also be mentioned in the field documentation).

File: verilog/analysis/verilog_linter_configuration.h [hunk 6]
```diff
@@ -130,6 +130,30 @@ struct ProjectPolicy {
   std::string ListPathGlobs() const;
 };
 
+struct LinterConfigurationFlags {
```
Comment (hzeller) [inline verilog/analysis/verilog_linter_configuration.h:133]: This name sounds very specific to be tied to flags. But in general this is more like a configuration struct, which we commonly would call somethinsomethingOptions; `LinterOptions` maybe ? ('Configuration' could be added but would not add more to the meaning).

File: verilog/analysis/verilog_linter_configuration.cc [hunk 7]
```diff
@@ -282,6 +283,71 @@ bool LinterConfiguration::operator==(const LinterConfiguration& config) const {
...
+    absl::string_view config_filename) {
+  // Read local configuration file
+  std::string content;
+
+  const absl::Status config_read_status =
+      verible::file::GetContents(config_filename, &content);
+  if (config_read_status.ok()) {
+    RuleBundle local_rules_bundle;
+    std::string error;
+    if (local_rules_bundle.ParseConfiguration(content, '\n', &error)) {
```
Comment (hzeller) [inline verilog/analysis/verilog_linter_configuration.cc:296]: It feels like we have done the 'read config into string, then send it to ParseConfiguration' somewhere else already ? Maybe we should have RuleBundle implement ParseConfigurationFile()...
(might be something for a later refactoring; if so, add TODO)

File: verilog/tools/lint/verilog_lint.cc [hunk 8]
```diff
@@ -74,9 +91,24 @@ int main(int argc, char** argv) {
   // All positional arguments are file names.  Exclude program name.
   for (const auto filename :
        verible::make_range(args.begin() + 1, args.end())) {
-    // Copy configuration, so that it can be locally modified per file.
-    const LinterConfiguration config(
-        verilog::LinterConfigurationFromFlags(filename));
+    // local configuration, so that it can be locally modified per file.
+    LinterConfiguration config;
+
+    const verilog::LinterConfigurationFlags flags = {
```
Comment (hzeller) [inline verilog/tools/lint/verilog_lint.cc:97]: So, personally, I'd probably solve this differently: keep the flags wherever they were before, but define the LinterConfigurationFlags struct (or, as suggested LinterConfigurationOptions).
The default constructore of LinterConfigurationOptions would just set the values all from the flags, so that you just would then assign values in main() that you'd like to be overriden.

That way, we keep the encapsulation of flags and keep the use of flags consistent in case we use juse the linter library in another tool. But it also allows for full unit-testing, as all the Options can be set manually.

File: verilog/analysis/verilog_linter.cc [hunk 9]
```diff
@@ -51,23 +50,6 @@
 #include "verilog/parser/verilog_token_classifications.h"
 #include "verilog/parser/verilog_token_enum.h"
 
-// TODO(hzeller): make --rules repeatable and cumulative
-ABSL_FLAG(verilog::RuleBundle, rules, {},
```
Comment (hzeller) [inline verilog/analysis/verilog_linter.cc:55]: I'd leave the flags here or in whatever compilation unit we define the constructor of LlinterConfigurationFlags (or, as I'd like to call it LinterConfigurationOptions). See comments below.

File: verilog/analysis/verilog_linter_configuration_test.cc [hunk 10]
```diff
@@ -775,5 +775,76 @@ TEST(RuleBundleTest, ParseRuleBundleSkipComments) {
...
+  LinterConfiguration config;
+
+  LinterConfigurationFlags flags = {
+      RuleSet::kAll,
+      RuleBundle(),
+      "-",
+      false,
+      false,
+      "filename"
+      "filename"};
```
Comment (hzeller) [inline verilog/analysis/verilog_linter_configuration_test.cc:805]: So a lot of unnamed parameters are a little bit hard to follow as we don't see the definition of LinterConfigurationFlags side-by-side. If the structs would only contain 2-3 parameters, {}-initiaalization would be fine.

Until C++20 catches up with C-89 and allows designated initializers (C++, why oh why arent they there?), we probably have to do manual init:
```
  flags.config_file = "-"
  flags.is_custom = false;
  // ...
```

I suspect that will also clash with the use of const references in the struct, but discussed above.

File: verilog/tools/lint/verilog_lint.cc (hunk 8 above)
Comment (tgorochowik) [reply verilog/tools/lint/verilog_lint.cc:97]: Sure, I can move them back. I moved them here as the official guidelines ([Flags Best Practices](https://abseil.io/docs/cpp/guides/flags)) say:

> * Prefer to define flags only in the file containing the binary’s main() function. Although Abseil flags may be defined anywhere in any namespace, avoid any usage outside of main() as it will otherwise be difficult to resolve conflicts.
> * Prefer to reference flags only from within the file containing the binary’s main() function, for the same reason.

And I thought that's what @fangism meant by cleaning the flags up - these are just guidelines however so I guess we can use a mix of both approaches.

File: verilog/analysis/verilog_linter_configuration.cc (hunk 7 above)
Comment (tgorochowik) [reply verilog/analysis/verilog_linter_configuration.cc:296]: I think this is the exact code you remember, I just moved it to a different file

File: verilog/analysis/verilog_linter_configuration.h (hunk 6 above)
Comment (tgorochowik) [reply verilog/analysis/verilog_linter_configuration.h]: Sure, makes sense, renamed

File: verilog/analysis/verilog_linter_configuration.h (hunk 5 above)
Comment (tgorochowik) [reply verilog/analysis/verilog_linter_configuration.h]: Sure, thanks, changed to a copy

File: verilog/analysis/verilog_linter_configuration.h (hunk 4 above)
Comment (tgorochowik) [reply verilog/analysis/verilog_linter_configuration.h]: Right, changed to strings

File: verilog/analysis/verilog_linter.cc (hunk 9 above)
Comment (tgorochowik) [reply verilog/analysis/verilog_linter.cc]: Moved the flags back

File: verilog/analysis/verilog_linter_configuration_test.cc (hunk 10 above)
Comment (tgorochowik) [reply verilog/analysis/verilog_linter_configuration_test.cc]: I think in our case we can use the designated initializes as we are initializing everything anyway. Thanks for the tip!

File: verilog/tools/lint/verilog_lint.cc (hunk 8 above)
Comment (fangism) [reply verilog/tools/lint/verilog_lint.cc]: Sorry, it was my idea to move them to main(), but they may still remain in the library, but be referenced from main().

File: verilog/analysis/verilog_linter_configuration_test.cc (hunk 10 above)
Comment (fangism) [reply verilog/analysis/verilog_linter_configuration_test.cc]: Keep in mind, the current minimum standard for building is `-std=c++11`.

File: verilog/tools/lint/verilog_lint.cc (hunk 8 above)
Comment (hzeller) [reply verilog/tools/lint/verilog_lint.cc]: ah ok. I think it is better for them to be where the struct is defined to set some defaults.

The ABSL guidelines are good, but it is also good to ignore them when it makes more sense :D
